Content.Server/Station/Systems/StationRespawnSystem.cs

```diff
@@ -0,0 +1,167 @@
+﻿using System.Linq;
+using Content.Server.Chat.Managers;
+using Content.Server.GameTicking;
+using Content.Server.Popups;
+using Content.Server.Station.Components;
+using Content.Shared.Maps;
+using Robust.Shared.Map;
+using Robust.Shared.Player;
+using Robust.Shared.Random;
+
+namespace Content.Server.Station.Systems;
+
+/// <summary>
+/// Handles respawning objects with StationRespawnComponent that get too far from the station, or get deleted.
+/// </summary>
+public sealed class StationRespawnSystem : EntitySystem
+{
+    [Dependency] private readonly StationSystem _stationSystem = default!;
+    [Dependency] private readonly EntityLookupSystem _lookupSystem = default!;
+    [Dependency] private readonly PopupSystem _popupSystem = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly IChatManager _chatManager = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly ITileDefinitionManager _tileDefMan = default!;
+
+    [ViewVariables(VVAccess.ReadWrite)]
+    public int MaximumRespawnAttempts { get; private set; } = 64;
+
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<StationRespawnComponent, MapInitEvent>(OnRespawnableSetup);
+        SubscribeLocalEvent<StationRespawnComponent, ComponentShutdown>(OnRespawnableDeletion);
+        SubscribeLocalEvent<StationRespawnControllerComponent, StationGridAddedEvent>(OnStationGridAdded);
+        SubscribeLocalEvent<StationInitializedEvent>(OnStationInitialized);
+    }
+
+    public override void Update(float frameTime)
+    {
+        foreach (var (respawn, xform) in EntityQuery<StationRespawnComponent, TransformComponent>())
+        {
+            if (respawn.AttachedStation is null || Deleted(respawn.AttachedStation))
+                continue;
+
+            if (Comp<StationDataComponent>(respawn.AttachedStation.Value).Grids.Count == 0)
+                continue;
+
+            var tooFar = true;
+            foreach (var grid in Comp<StationDataComponent>(respawn.AttachedStation.Value).Grids)
+            {
+                if (Deleted(grid))
+                    continue; //
+                var distance = (Transform(grid).WorldPosition - xform.WorldPosition).Length;
```

Comment:
You should cache the transform's worldpos first before iterating as it's very expensive.